trinity/config.py

```diff
@@ -440,14 +444,77 @@ def nodekey(self, value: Union[bytes, PrivateKey]) -> None:
                 f"`PrivateKey` instance: got {type(self._nodekey)}"
             )
 
+    @contextmanager
+    def process_id_file(self, process_name: str):  # type: ignore
+        with PidFile(process_name, self.data_dir):
+            yield
+
     @classmethod
-    def from_parser_args(cls, parser_args: argparse.Namespace) -> 'TrinityConfig':
+    def from_parser_args(cls,
+                         parser_args: argparse.Namespace,
+                         app_config_types: Iterable[Type['BaseAppConfig']]) -> 'TrinityConfig':
         """
         Helper function for initializing from the namespace object produced by
         an ``argparse.ArgumentParser``
         """
         constructor_kwargs = construct_trinity_config_params(parser_args)
-        return cls(**constructor_kwargs)
+        trinity_config = cls(**constructor_kwargs)
+
+        trinity_config.initialize_app_configs(parser_args, app_config_types)
+
+        return trinity_config
+
+    def initialize_app_configs(self,
+                               parser_args: argparse.Namespace,
+                               app_config_types: Iterable[Type['BaseAppConfig']]) -> None:
+
+        for app_config_type in app_config_types:
+            self.add_app_config(app_config_type.from_parser_args(parser_args, self))
+
+    def add_app_config(self, app_config: 'BaseAppConfig') -> None:
+        self._app_configs[type(app_config)] = app_config
+
+    def has_app_config(self, app_config_type: Type['BaseAppConfig']) -> bool:
+        return app_config_type in self._app_configs.keys()
+
+    def get_app_config(self, app_config_type: Type[TAppConfig]) -> TAppConfig:
+        # We want this API to return the specific type of the app config that is requested.
+        # Our backing field only knows that it is holding `BaseAppConfig`'s but not concrete types
+        return cast(TAppConfig, self._app_configs[app_config_type])
+
+
+class BaseAppConfig(ABC):
+
+    @classmethod
+    @abstractmethod
+    def from_parser_args(cls,
+                         args: argparse.Namespace,
+                         base_config: TrinityConfig) -> 'BaseAppConfig':
+        pass
+
+
+class DefaultBaseAppConfig(BaseAppConfig):
```

Comment:
@pipermerriam would this be an acceptable compromise? From the perspective of the `TrinityConfig` there exists only `BaseAppConfig` as an abstraction and it only dictates the `from_parser_args` class method.

However, additionally we introduce a `DefaultBaseAppConfig` as a convenience abstraction that dictates the `args` / `base_config` convention. So we maintain the full flexibility from the perspective of the `TrinityConfig` but standardize the common scheme.

Reply:
I'm leaning towards :-1: 

- I don't see a clear benefit (if it isn't truly standardized then it's not standard)
- it's an attempt at establishing an abstraction when I don't think we really *know* what the abstraction looks like...
- it results in the `__init__` being tied to an `argparse.Namespace` shaped object which likely makes testing harder.

Thoughts?

Reply:
>I don't see a clear benefit

The benefit would be not having to implement `from_parser_args` and `__init__` in every `AppConfig`

>it's an attempt at establishing an abstraction when I don't think we really know what the abstraction looks like...

Mmmh...I think most `AppConfig` implementations will just take the `args` and the `trinity_config` in their `__init__` anyway or what are the cases where you see them taking other things (especially considering, that they all origin in a top level `from_parser_args` call from the `TrinityConfig`

>t results in the `__init__` being tied to an `argparse.Namespace` shaped object which likely makes testing harder.

Well, without this abstraction, we will just have the exact same `__init__` with ` argsparse.Namespace` on every implementation of `AppConfig`

Reply:
> I think most `AppConfig` implementations will just take the `args` and the `trinity_config` in their `__init__` anyway

I think this suggests otherwise:

https://github.com/ethereum/py-evm/blob/751c856953060d68ae7be2c62306d11935e32345/trinity/utils/chains.py#L123-L163

In the case of  `__init__(args, trinity_config)` the result is that the `__init__` method must parse and normalize the values from the `argparse.Namespace` object which I would argue is a bad coupling,  binding it to the CLI

In the case of `__init__(trinity_config, arg_1, arg_2, ...)` and `Config.from_parser_args(args, trinity_config)` we still have *coupling* but it is looser and keeps the `cli_args -> init_args` conversion from bleeding into the actual configuration objects implementation.

A **lot** of my reasoning is centered around how the `Config` classes are tested so admittedly this doesn't have much effect on the end result as far as production code goes.

I'm :+1: on whatever you decide.

Reply:
>A lot of my reasoning is centered around how the Config classes are tested so admittedly this doesn't have much effect on the end result as far as production code goes.

Ah, I missed that goal to spread out the `args` into individual parameters for better testing. That sounds reasonable.

I would however, still favor to at least provide an `__init__(trinity_config: TrinityConfig)__` and have child classes overwrite and extend that. Not having the `trinity_config` access standardized just doesn't sit so well with me.